pandas/core/nanops.py

```diff
@@ -254,9 +254,16 @@ def nansum(values, axis=None, skipna=True):
 @bottleneck_switch()
 def nanmean(values, axis=None, skipna=True):
     values, mask, dtype, dtype_max = _get_values(values, skipna, 0)
-    the_sum = _ensure_numeric(values.sum(axis, dtype=dtype_max))
     count = _get_counts(mask, axis)
 
+    dtype_sum = dtype_max
+    if is_integer_dtype(dtype):
+        dtype_sum = np.float64
+    elif is_float_dtype(dtype):
+        dtype_sum = dtype
+        count = dtype.type(count)
```

Comment:
@jreback actually changing count to `int64` may not help, this is what I'm seeing:

```
In [1]: import numpy as np

In [2]: (np.float32(100) / np.int64(100)).dtype
Out[2]: dtype('float64')

In [3]: (np.float32(100) / np.int32(100)).dtype
Out[3]: dtype('float64')

In [4]: (np.float32(100) / np.float32(100)).dtype
Out[4]: dtype('float32')
```

It still gets cast to `float64` unless we have the denominator as `float32` as well.